test/tls.c

```diff
@@ -196,9 +208,11 @@ check_tls(char *where, bool check_addr, void *tls_region)
 	check_inside_tls_region(&bss_var, tls_region);
 	check_inside_tls_region(&overaligned_bss_var, tls_region);
 
-        if (__non_tls_bss_start < __tdata_start + _tls_size() || __tdata_start + _tls_size() + 64 < __non_tls_bss_start) {
-                printf("non-TLS bss data doesn't start after TLS data (is %p should be %p)\n",
-                       __non_tls_bss_start, __tdata_start + _tls_size());
+        char* tdata_end = __tdata_start + _tls_size();
+        char* non_tls_bss_start_latest = __align_up(tdata_end + 64, OVERALIGN_NON_TLS_BSS);
```

Comment:
(thanks for creating some local variables here to make it easier to reason about).

I think you want to bound the value of `__non_tls_bss_start` to be within a short distance of the end of the TLS data, after that has been aligned:
```
char *non_tls_bss_start_latest = __align_up(tdata_end, OVERALIGN_NON_TLS_BSS) + 64;
if (__non_tls_bss_start < tdata_end || non_tls_bss_start_latest < __non_tls_bss_start)
```

Reply:
hrm. that's not quite right either -- it's possible that the non-tls BSS data will end up further away than that, if the end of TLS data hits right at an aligned address and there's another chunk of BSS before the aligned variable causing that to get pushed another alignment interval out. Maybe just make the allowed variance be `OVERALIGN_NON_TLS_BSS + 64` instead of `64`?

Reply:
Hopefully the new version makes sense?